pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/refreshsegment/RefreshSegmentTaskExecutor.java

```diff
@@ -0,0 +1,203 @@
+    String taskType = pinotTaskConfig.getTaskType();
+
+    LOGGER.info("Starting task: {} with configs: {}", taskType, configs);
+
+    TableConfig tableConfig = getTableConfig(tableNameWithType);
+    Schema schema = getSchema(tableNameWithType);
+
+    IndexLoadingConfig indexLoadingConfig = new IndexLoadingConfig(tableConfig, schema);
+    SegmentMetadataImpl segmentMetadata = new SegmentMetadataImpl(indexDir);
+    PinotConfiguration segmentDirectoryConfigs = indexLoadingConfig.getSegmentDirectoryConfigs();
+    SegmentDirectoryLoaderContext segmentLoaderContext =
+        new SegmentDirectoryLoaderContext.Builder().setTableConfig(indexLoadingConfig.getTableConfig())
+            .setSchema(schema)
+            .setInstanceId(indexLoadingConfig.getInstanceId())
+            .setSegmentName(segmentMetadata.getName())
+            .setSegmentCrc(segmentMetadata.getCrc())
+            .setSegmentDirectoryConfigs(segmentDirectoryConfigs)
+            .build();
+    SegmentDirectory segmentDirectory =
+        SegmentDirectoryLoaderRegistry.getDefaultSegmentDirectoryLoader().load(indexDir.toURI(), segmentLoaderContext);
+
+    // BaseDefaultColumnHandler part of needPreprocess() does not process any changes to existing columns like datatype,
+    // change from dimension to metric, etc.
+    boolean needPreprocess = ImmutableSegmentLoader.needPreprocess(segmentDirectory, indexLoadingConfig, schema);
+    closeSegmentDirectoryQuietly(segmentDirectory);
+    Set<String> refreshColumnSet = new HashSet<>();
+
+    for (FieldSpec fieldSpecInSchema : schema.getAllFieldSpecs()) {
+      // Virtual columns are constructed while loading the segment, thus do not exist in the record, nor should be
+      // persisted to the disk.
+      if (fieldSpecInSchema.isVirtualColumn()) {
+        continue;
+      }
+
+      String column = fieldSpecInSchema.getName();
+      ColumnMetadata columnMetadata = segmentMetadata.getColumnMetadataFor(column);
+      if (columnMetadata != null) {
+        FieldSpec fieldSpecInSegment = columnMetadata.getFieldSpec();
+
+        // Check the data type and default value matches.
+        FieldSpec.DataType dataTypeInSegment = fieldSpecInSegment.getDataType();
+        FieldSpec.DataType dataTypeInSchema = fieldSpecInSchema.getDataType();
+
+        // Column exists in segment.
+        if (dataTypeInSegment != dataTypeInSchema) {
+          // Check if we need to update the data-type. DataType change is dependent on segmentGeneration code converting
+          // the object to the destination datatype. If the existing data is the column is not compatible with the
+          // destination data-type, the refresh task will fail.
+          refreshColumnSet.add(column);
+        }
+
+        // TODO: Maybe we can support singleValue to multi-value conversions are supproted and vice-versa.
+      } else {
+        refreshColumnSet.add(column);
+      }
+    }
+
+    if (!needPreprocess && refreshColumnSet.isEmpty()) {
```

Comment:
This is the no-op case where we dont need to do anything to that segment, right? 
if thats the case, why do we need to upload the segment which happens in **BaseSingleSegmentConversionExecutor** and also update Segment Zk metadata. 

try {
        SegmentConversionUtils.uploadSegment(configs, httpHeaders, parameters, tableNameWithType, segmentName,
            uploadURL, convertedTarredSegmentFile);
      } 

I feel we should not go through this code path unnecessarily when the segment requires no update/refresh.

Reply:
Good catch. I have to update the code to set file name. This updates the last processed time and uploads the segment. The upload is purely a [ZK metadata update as the CRCs will match](https://github.com/apache/pinot/blob/88f4916eee45bc150fdbef7a816408a2bd61413b/pinot-controller/src/main/java/org/apache/pinot/controller/api/upload/ZKOperator.java#L258)